lib/enmeshed/connector_api.rb

```diff
@@ -0,0 +1,244 @@
+      parse_result(response).map {|relationship_json| Relationship.new(relationship_json) }
+    end
+    private_class_method :parse_relationships
+
+    def self.relationship_template(nbp_uid)
+      {
+        maxNumberOfAllocations: 1,
+        expiresAt: TEMPLATE_VALIDITY_PERIOD.from_now,
+        content: {
+          metadata: {nbp_uid:},
+          '@type': 'RelationshipTemplateContent',
+          onNewRelationship: {
+            '@type': 'Request',
+            items: [
+              {
+                '@type': 'ShareAttributeRequestItem',
+                mustBeAccepted: true,
+                attribute: display_name_attribute,
+                sourceAttributeId: display_name_id,
+              },
+            ] + required_attributes_template,
+          },
+        },
+      }
+    end
+    private_class_method :relationship_template
+
+    def self.required_attributes_template
+      REQUIRED_ATTRIBUTES.map do |attr|
+        {
+          '@type': 'ReadAttributeRequestItem',
+          mustBeAccepted: true,
+          query: {
+            '@type': 'IdentityAttributeQuery',
+            valueType: attr,
+          },
+        }
+      end
+    end
+    private_class_method :required_attributes_template
+
+    def self.display_name_attribute
+      {
+        '@type': 'IdentityAttribute',
+        owner: enmeshed_address,
+        value: {
+          '@type': 'DisplayName',
+          value: DISPLAY_NAME,
+        },
+      }
+    end
+    private_class_method :display_name_attribute
+
+    def self.conn
+      @conn ||= init_conn
+    end
+    private_class_method :conn
+
+    def self.init_conn
+      if User.omniauth_providers.exclude?(:nbp) || CONNECTOR_URL.nil? || API_KEY.nil? || DISPLAY_NAME.nil?
+        raise ConnectorError.new('NBP provider or enmeshed connector not configured as expected')
+      end
+
+      Faraday.new(CONNECTOR_URL, headers:)
+    end
+    private_class_method :init_conn
+
+    def self.enmeshed_address
+      return @enmeshed_address if @enmeshed_address.present?
+
+      identity = parse_result conn.get('/api/v2/Account/IdentityInfo')
+      @enmeshed_address = identity[:address]
+    end
+    private_class_method :enmeshed_address
+
+    def self.display_name_id
+      return @display_name_id if @display_name_id.present?
+
+      new_attribute = parse_result(conn.post('/api/v2/Attributes') do |req|
+        req.body = {content: display_name_attribute}.to_json
+      end)
+
+      @display_name_id = new_attribute[:id]
+    end
+    private_class_method :display_name_id
+
+    def self.headers
+      {'X-API-KEY': API_KEY, 'content-type': 'application/json', accept: 'application/json'}
+    end
+    private_class_method :headers
+  end
+
+  class Relationship
+    def initialize(relationship_json)
+      @json = relationship_json
+      @relationship_changes = relationship_json[:changes] || []
+    end
+
+    attr_reader :json
+
+    def userdata
+      @userdata || parse_userdata
```

Comment:
One could unify this, too, but we don't mind (I have a very slight tendency towards unification, but really not a strong one):

```suggestion
      @userdata ||= parse_userdata
```

Of course, this would also allow to omit the assignment in the `parse_userdata`.